src/activity_actor.cpp

```diff
@@ -1124,36 +1124,101 @@ std::unique_ptr<activity_actor> hotwire_car_activity_actor::deserialize( JsonVal
+                                        const cata::value_ptr<activity_data_common> map_data_common_t::* const tool_member, bool finish,
+                                        activity_data_common *&data )
 {
-    const map &here = get_map();
-    int moves_before_quality;
-
     if( here.has_furn( target ) ) {
         const furn_id &furn_type = here.furn( target );
-        if( !furn_type->hacksaw->valid() ) {
+        const auto furn_tool_ad = ( *furn_type.*tool_member );
+        if( !furn_tool_ad->valid() ) {
             if( !testing ) {
-                debugmsg( "%s hacksaw is invalid", furn_type.id().str() );
+                debugmsg( "%s %s is invalid", furn_type.id().str(), tool_name );
             }
             act.set_to_null();
-            return;
+            return false;
         }
 
-        moves_before_quality = to_moves<int>( furn_type->hacksaw->duration() );
+        if( finish ) {
+            const furn_str_id new_furn = std::get<furn_str_id>( furn_tool_ad->result() );
+            if( !new_furn.is_valid() ) {
+                if( !testing ) {
+                    debugmsg( "%s furniture: %s invalid furniture", tool_name, new_furn.str() );
+                }
+                act.set_to_null();
+                return false;
+            }
+
+            data = &*furn_tool_ad;
+            here.furn_set( target, new_furn );
+        } else {
+            moves = to_moves<int>( ( *furn_type.*tool_member )->duration() );
+        }
     } else if( const ter_id &ter_type = here.ter( target ); !ter_type->is_null() ) {
-        if( !ter_type->hacksaw->valid() ) {
+        const auto ter_tool_ad = ( *ter_type.*tool_member );
+        if( !ter_tool_ad->valid() ) {
             if( !testing ) {
-                debugmsg( "%s hacksaw is invalid", ter_type.id().str() );
+                debugmsg( "%s %s is invalid", ter_type.id().str(), tool_name );
             }
             act.set_to_null();
-            return;
+            return false;
+        }
+        if( finish ) {
+            const ter_str_id new_ter = std::get<ter_str_id>( ter_tool_ad->result() );
+            if( !new_ter.is_valid() ) {
+                if( !testing ) {
+                    debugmsg( "%s terrain: %s invalid terrain", tool_name, new_ter.str() );
+                }
+                act.set_to_null();
+                return false;
+            }
+
+            data = &*ter_tool_ad;
+            here.ter_set( target, new_ter );
+        } else {
+            moves = to_moves<int>( ter_tool_ad->duration() );
         }
-        moves_before_quality = to_moves<int>( ter_type->hacksaw->duration() );
     } else {
         if( !testing ) {
-            debugmsg( "hacksaw activity called on invalid terrain" );
+            debugmsg( "%s activity called on invalid terrain", tool_name );
         }
         act.set_to_null();
+        return false;
+    }
+
+    if( finish ) {
```

Comment:
Instead of these conditional blocks to differentiate start from finish, I could make this function take function pointer params, then have separate `start` and `finish` functions for those params.